bytecomp/matching.ml

```diff
@@ -430,76 +430,52 @@ end = struct
-          | Tpat_variant (lab', Some arg, _) when lab = lab' -> (p, arg :: rem)
-          | Tpat_any -> (p, omega :: rem)
-          | _ -> raise NoMatch
-      )
-    | Tpat_variant (lab, None, _) -> (
-        fun q rem ->
-          match q.pat_desc with
-          | Tpat_variant (lab', None, _) when lab = lab' -> (p, rem)
-          | Tpat_any -> (p, rem)
-          | _ -> raise NoMatch
-      )
-    | Tpat_array omegas -> (
-        let len = List.length omegas in
-        fun q rem ->
-          match q.pat_desc with
-          | Tpat_array args when List.length args = len -> (p, args @ rem)
-          | Tpat_any -> (p, omegas @ rem)
-          | _ -> raise NoMatch
-      )
-    | Tpat_tuple omegas -> (
-        let len = List.length omegas in
-        fun q rem ->
-          match q.pat_desc with
-          | Tpat_tuple args when List.length args = len -> (p, args @ rem)
-          | Tpat_any -> (p, omegas @ rem)
-          | _ -> raise NoMatch
-      )
-    | Tpat_record (((_, lbl, _) :: _ as l), _) -> (
-        (* Records are normalized *)
-        let len = Array.length lbl.lbl_all in
-        fun q rem ->
-          match q.pat_desc with
-          | Tpat_record (((_, lbl', _) :: _ as l'), _)
-            when Array.length lbl'.lbl_all = len ->
-              let l' = all_record_args l' in
-              (p, List.fold_right (fun (_, _, p) r -> p :: r) l' rem)
-          | Tpat_any -> (p, List.fold_right (fun (_, _, p) r -> p :: r) l rem)
-          | _ -> raise NoMatch
-      )
-    | Tpat_lazy omega -> (
-        fun q rem ->
-          match q.pat_desc with
-          | Tpat_lazy arg -> (p, arg :: rem)
-          | Tpat_any -> (p, omega :: rem)
-          | _ -> raise NoMatch
-      )
-    | _ -> fatal_error "Matching.Context.matcher"
+  let ctx_matcher p q rem =
+    let rec expand_record p =
+      match p.pat_desc with
+        | Tpat_record (l, _) ->
+           {p with pat_desc = Tpat_record (all_record_args l, Closed)}
+        | Tpat_alias (p, _, _) -> expand_record p
+        | _ -> p in
+    let ph, omegas =
+      let ph, p_args = Pattern_head.deconstruct (expand_record p) in
+      ph, List.map (fun _ -> omega) p_args in
+    let qh, args = Pattern_head.deconstruct (expand_record q) in
+    let yes () = (p, args @ rem) in
+    let no () = raise NoMatch in
+    let yesif b = if b then yes () else no () in
+    match Pattern_head.desc ph, Pattern_head.desc qh with
+      | Any, _ -> fatal_error "Matching.Context.matcher"
+      | _, Any -> (p, omegas @ rem)
+
+      | Construct cstr, Construct cstr' ->
+         (* NB: may_equal_constr considers (potential) constructor rebinding *)
+         yesif (Types.may_equal_constr cstr cstr')
+      | Construct _, _ -> no ()
+
+      | Constant cst, Constant cst' ->
+         yesif (const_compare cst cst' = 0)
+      | Constant _, _ -> no ()
+
+      | Variant { tag; _ }, Variant { tag = tag'; _ } ->
+         yesif (tag = tag')
+      | Variant _, _ -> no ()
```

Comment:
There is a difference here, if you have two variants with the same name/tag but one of them has an arg and not the other, then you will still `yes ()`, which is wrong.

Reply:
Fixed, thanks.